fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBMapQueryTest.java

```diff
@@ -0,0 +1,1241 @@
+                            .build());
+                }
+                assertEquals(outerRecord.getMap().getEntryList(), evaluatedEntries);
+            }
+
+            commit(context);
+        }
+    }
+
+    @Test
+    void filterKeyAndProjectValueOnUnnestedMinimalIndex() {
+        RecordMetaDataHook hook = addUnnestedType()
+                .andThen(metaDataBuilder -> metaDataBuilder.addIndex(OUTER_WITH_ENTRIES, unnestedKeyAndValue()));
+        filterKeyAndProjectValueOnUnnested(hook,
+                coveringIndexPlan()
+                    .where(indexPlanOf(
+                            indexPlan()
+                                    .where(indexName(UNNESTED_KEY_AND_VALUE))
+                                    .and(scanComparisons(range("[EQUALS $key]")))
+                    ))
+        );
+    }
+
+    @Test
+    void filterKeyAndProjectValueUnnestedWithOtherFieldInIndex() {
+        RecordMetaDataHook hook = addUnnestedType()
+                .andThen(metaDataBuilder -> metaDataBuilder.addIndex(OUTER_WITH_ENTRIES, unnestedKeyOtherValue()));
+        filterKeyAndProjectValueOnUnnested(hook,
+                coveringIndexPlan()
+                        .where(indexPlanOf(
+                                indexPlan()
+                                        .where(indexName(UNNESTED_KEY_OTHER_VALUE))
+                                        .and(scanComparisons(range("[EQUALS $key]")))
+                        ))
+        );
+    }
+
+    private void filterKeyAndProjectValueOnUnnested(RecordMetaDataHook hook, BindingMatcher<? extends RecordQueryPlan> planMatcher) {
+        List<TestRecordsNestedMapProto.OuterRecord> data = setUpData(hook);
+        Set<String> keys = mapKeys(data);
+        assertThat(keys, not(empty()));
+
+        try (FDBRecordContext context = openContext()) {
+            createOrOpenMapStore(context, hook);
+
+            final String keyParam = "key";
+            RecordQuery query = RecordQuery.newBuilder()
+                    .setRecordType(OUTER_WITH_ENTRIES)
+                    .setFilter(Query.field("entry").matches(Query.field("key").equalsParameter(keyParam)))
+                    .setRequiredResults(List.of(field("entry").nest("value")))
+                    .build();
+            RecordQueryPlan plan = planner.plan(query);
+            assertMatchesExactly(plan, planMatcher);
+
+            final KeyExpression recIdExpr = field(UnnestedRecordType.PARENT_CONSTITUENT).nest("rec_id");
+            final KeyExpression keyAndValueExpr = field("entry").nest(concatenateFields("key", "value"));
+            for (String key : keys) {
+                EvaluationContext evaluationContext = EvaluationContext.forBinding(keyParam, key);
+                final List<TestRecordsNestedMapProto.OuterRecord> queried = plan.execute(recordStore, evaluationContext)
+                        .map(rec -> {
+                            // Nest the data back into an OuterRecord
```

Comment:
It's probably not worth changing, but for future reference stylistic advice, I found this hard to follow. It's not clear why you map it back into an `OuterRecord` rather than simply asserting directly about the synthetic record that comes back.